src/cc/libbpf.c

```diff
@@ -796,15 +797,18 @@ static int bpf_try_perf_event_open_with_probe(const char *name, uint64_t offs,
   /* config1 here is kprobe_func or  uprobe_path */
   attr.config1 = ptr_to_u64((void *)name);
   // PID filter is only possible for uprobe events.
-  if (pid < 0)
+  if (cgroup_fd >= 0) {
+    flags |= PERF_FLAG_PID_CGROUP;
+    pid = cgroup_fd;
+  } else if (pid < 0)
     pid = -1;
   // perf_event_open API doesn't allow both pid and cpu to be -1.
   // So only set it to -1 when PID is not -1.
   // Tracing events do not do CPU filtering in any cases.
-  if (pid != -1)
+  if (pid != -1 && !(flags & PERF_FLAG_PID_CGROUP))
```

Comment:
Why `!(flags & PERF_FLAG_PID_CGROUP)`?

Reply:
> I feel we shouldn't change the general kprobe API for this specific case. There are many more controls in the `perf_event_open` and we probably couldn't add all of them to the kprobe API.

As long as we keep compatibility to existing code, I don't see there would be problems to extend kprobe API. Profiling performance in container wise is a very important feature for cluster.

Reply:
> Why `!(flags & PERF_FLAG_PID_CGROUP)`?

Kernel doesn't allow coexistence of `cpu` to be `-1` and this flag.

Reply:
Does anyone know why kernel doesn't allow to have cpu and pid setting to `-1` at the same time? Perf tools have to work this around by creating many per-cpu task for system wide tracing.
 
For kprobe, kernel doesn't seems to have implemented CPU filtering at this point, so bcc tools are taking this advantage. I wonder if kernel has got this `fixed`, would it break all existing tools? Please let me know if I understood it wrong.

Reply:
Maybe some comments are better here.

Reply:
I think a lot of these are just general parameter checks from the `perf_event_open` syscall and it would be hard to make exception for BPF events. I don't think CPU filtering matter with kprobe / uprobe / Kernel Tracepoints, so let's just keep it this way.